File: src/rules/checks/reasoning.ts
// Reasoning rules: AGUI401–AGUI402, plus REASONING_MESSAGE_CHUNK handling.
//
// Per SQ-4: block-level nesting (REASONING_MESSAGE_* inside
// REASONING_START/END) is only described as "a typical flow" by the docs, so
// it is NOT enforced. AGUI401 enforces the message-level pairing, mirroring
// the text message rules; AGUI402 reports unterminated starts at warning.

import type { CheckApi, RunState, EmitFn } from "./context.js"
import { str } from "./context.js"

export function handleReasoningEvent(api: CheckApi): void {
  const { type, event, run, emit, index } = api

  switch (type) {
    case "REASONING_START": {
      const id = str(event, "messageId")
      if (id !== undefined) run.openReasoningBlocks.set(id, index)
      return
    }

    case "REASONING_END": {
      const id = str(event, "messageId")
      if (id !== undefined) run.openReasoningBlocks.delete(id)
      return
    }

    case "REASONING_MESSAGE_START": {
      const id = str(event, "messageId")
      if (id !== undefined) run.openReasoningMessages.set(id, index)
      return
    }

    case "REASONING_MESSAGE_CONTENT": {
      const id = str(event, "messageId")
      if (id === undefined) return
      const openChunk = run.reasoningChunk !== null && run.reasoningChunk.messageId === id
      if (!run.openReasoningMessages.has(id) && !openChunk) {
        emit("AGUI401", { messageId: id }, { pointer: "/messageId" })
      }
      return
    }

    case "REASONING_MESSAGE_END": {
      const id = str(event, "messageId")
      if (id !== undefined) run.openReasoningMessages.delete(id)
      return
    }

    case "REASONING_MESSAGE_CHUNK": {
      const id = str(event, "messageId")
      if (id === undefined) {
        if (run.reasoningChunk === null) {
          emit("AGUI504", { type, detail: "first REASONING_MESSAGE_CHUNK must include messageId" }, {
            pointer: "/messageId",
          })
          return
        }
        if (event.delta === "") run.reasoningChunk = null // documented implicit close
        return
      }
      if (run.reasoningChunk !== null && run.reasoningChunk.messageId !== id) {
        run.reasoningChunk = null
      }
      if (event.delta === "") {
        run.reasoningChunk = null // empty delta implicitly closes the message
        return
      }
      if (run.reasoningChunk === null) {
        run.reasoningChunk = { messageId: id, startIndex: index }
      }
      return
    }

    // REASONING_ENCRYPTED_VALUE: schema-checked only; no ordering rules cited.
  }
}

/** Chunked reasoning also closes on any non-reasoning event (documented). */
export function closeReasoningChunk(run: RunState): void {
  run.reasoningChunk = null
}

/** AGUI402 — unterminated reasoning at a clean run end. */
export function endOfRunReasoning(run: RunState, emit: EmitFn, atIndex: number): void {
  for (const [id, startIndex] of run.openReasoningBlocks) {
    emit("AGUI402", { startType: "REASONING_START", messageId: id }, {
      eventIndex: atIndex,
      relatedEventIndex: startIndex,
    })
  }
  for (const [id, startIndex] of run.openReasoningMessages) {
    emit("AGUI402", { startType: "REASONING_MESSAGE_START", messageId: id }, {
      eventIndex: atIndex,
      relatedEventIndex: startIndex,
    })
  }
}

File: py/src/ag_ui_validate/rules/checks/reasoning.py
"""Reasoning rules: AGUI401-AGUI402, plus REASONING_MESSAGE_CHUNK handling.

Per SQ-4: block-level nesting (REASONING_MESSAGE_* inside
REASONING_START/END) is only described as "a typical flow" by the docs, so
it is NOT enforced. AGUI401 enforces the message-level pairing, mirroring
the text message rules; AGUI402 reports unterminated starts at warning.
Mirrors js/src/rules/checks/reasoning.ts.
"""

from __future__ import annotations

from .context import CheckApi, EmitFn, ReasoningChunk, RunState, str_field


def handle_reasoning_event(api: CheckApi) -> None:
    run = api.run
    event = api.event
    index = api.index
    emit = api.emit
    type_ = api.type

    if type_ == "REASONING_START":
        message_id = str_field(event, "messageId")
        if message_id is not None:
            run.open_reasoning_blocks[message_id] = index
        return

    if type_ == "REASONING_END":
        message_id = str_field(event, "messageId")
        if message_id is not None:
            run.open_reasoning_blocks.pop(message_id, None)
        return

    if type_ == "REASONING_MESSAGE_START":
        message_id = str_field(event, "messageId")
        if message_id is not None:
            run.open_reasoning_messages[message_id] = index
        return

    if type_ == "REASONING_MESSAGE_CONTENT":
        message_id = str_field(event, "messageId")
        if message_id is None:
            return
        open_chunk = run.reasoning_chunk is not None and run.reasoning_chunk.message_id == message_id
        if message_id not in run.open_reasoning_messages and not open_chunk:
            emit("AGUI401", {"messageId": message_id}, {"pointer": "/messageId"})
        return

    if type_ == "REASONING_MESSAGE_END":
        message_id = str_field(event, "messageId")
        if message_id is not None:
            run.open_reasoning_messages.pop(message_id, None)
        return

    if type_ == "REASONING_MESSAGE_CHUNK":
        message_id = str_field(event, "messageId")
        if message_id is None:
            if run.reasoning_chunk is None:
                emit(
                    "AGUI504",
                    {"type": type_, "detail": "first REASONING_MESSAGE_CHUNK must include messageId"},
                    {"pointer": "/messageId"},
                )
                return
            if event.get("delta") == "":
                run.reasoning_chunk = None  # documented implicit close
            return
        if run.reasoning_chunk is not None and run.reasoning_chunk.message_id != message_id:
            run.reasoning_chunk = None
        if event.get("delta") == "":
            run.reasoning_chunk = None  # empty delta implicitly closes the message
            return
        if run.reasoning_chunk is None:
            run.reasoning_chunk = ReasoningChunk(message_id=message_id, start_index=index)
        return

    # REASONING_ENCRYPTED_VALUE: schema-checked only; no ordering rules cited.


def close_reasoning_chunk(run: RunState) -> None:
    """Chunked reasoning also closes on any non-reasoning event (documented)."""
    run.reasoning_chunk = None


def end_of_run_reasoning(run: RunState, emit: EmitFn, at_index: int) -> None:
    """AGUI402 - unterminated reasoning at a clean run end."""
    for message_id, start_index in run.open_reasoning_blocks.items():
        emit(
            "AGUI402",
            {"startType": "REASONING_START", "messageId": message_id},
            {"eventIndex": at_index, "relatedEventIndex": start_index},
        )
    for message_id, start_index in run.open_reasoning_messages.items():
        emit(
            "AGUI402",
            {"startType": "REASONING_MESSAGE_START", "messageId": message_id},
            {"eventIndex": at_index, "relatedEventIndex": start_index},
        )
